### ISSUE TITLE
Truncated ( ... ) download still happening even after #133 fix

### ISSUE THREAD

**Issue by mukuntharajaa** [lorenzodifuccia/safaribooks#150]
I am on master branch and currently updated to Oct 14 2019 commit. Still I am seeing truncated chapter downloads.

Book id: 9781491908419

Chapter 2: Item 5: second page shows "..." and Item 6 is altogether missing.

Please let me know, if any further information is required.

**Comment by brookscl** [comment]
Same here.

**Comment by vavdoshka** [comment]
+1

**Comment by spac-valentin** [comment]
+1

**Comment by phamhoangtuan** [comment]
+1

**Comment by sorinescu** [comment]
+1

**Comment by manfredlotz** [comment]
I have the same issue

**Comment by Azarakhsh** [comment]
+1

**Comment by ghistes** [comment]
I have the same problem.

It seems to me that the problem is the login. Even though you get a 200 response-code when logging in, you never get a sessionid-cookie, and for that reason when requesting the chapters you are treated as if you are not logged in, resulting in the truncations - at least that's how it looked to me when I was trying to understand what is going on (not sure if it helps...).

**Comment by elrob** [comment]
The above PR fixes this issue for me

**Comment by manfredlotz** [comment]
Your fix worked fine for me too. Thanks a lot for your work!

**Comment by manfredlotz** [comment]
I tried a couple of downloads, and mostly the epubs are not really usable. @elrob : However, this is not the fault of your fix.

**Comment by milktea02** [comment]
Still having issues :( even with #152

**Comment by elrob** [comment]
> I tried a couple of downloads, and mostly the epubs are not really usable. @elrob : However, this is not the fault of your fix.

This issue is about the truncation of output as if you're not logged in. The PR I created doesn't change anything in the epub creation. I have had no issues with three books I've since tested with. Definitely usable for me. Can you give me an example of a book you've had issues with? And what those issues are?

**Comment by elrob** [comment]
> Still having issues :( even with #152

@milktea02 What issues are you having? Are they related to truncation (this github issue tracks the truncation problem)?

**Comment by mukuntharajaa** [comment]
> > Still having issues :( even with #152
> 
> @milktea02 What issues are you having? Are they related to truncation (this github issue tracks the truncation problem)?

I have tried the same book again ( 9781491908419 ). I am able to see contents now without ellipsis. But when I click chapter 6, it takes me to last page of chapter 6 properly, but shows chapter 5 as highlighted on the left hand side layout.

Guess this is some minor stuff.

**Comment by elrob** [comment]
> 
> I have tried the same book again ( 9781491908419 ). I am able to see contents now without ellipsis. But when I click chapter 6, it takes me to last page of chapter 6 properly, but shows chapter 5 as highlighted on the left hand side layout.
> 
> Guess this is some minor stuff.

@mukuntharajaa Thanks for the response. If it is an issue you would like to raise and get fixed then I recommend creating a new github issue for it. This github issue was around the truncation of chapters due to authentication issues. So for now, if/when @lorenzodifuccia accepts #152 then github issue would be fixed.

**Comment by varta2014** [comment]
can we try this code please 
thank you

**Comment by elrob** [comment]
> can we try this code please
> thank you

@varta2014 If you want to try my change before it is merged into this repository then you can just pull it from https://github.com/elrob/safaribooks

**Comment by manfredlotz** [comment]
@elrob Unfortunately, I don't remember which book download I tried. I know that FBReader crashed when opening the epub. The last downloads I did were ok.

**Comment by milktea02** [comment]
> > Still having issues :( even with #152
> 
> @milktea02 What issues are you having? Are they related to truncation (this github issue tracks the truncation problem)?

Tried Clean Code (9780136083238) and still get truncation. I'm logging in via SSO if that might be the issue.

**Comment by AsimShakour** [comment]
I am having truncation with book: 9781119449270 in this area: https://learning.oreilly.com/library/view/professional-c-7/9781119449270/fintro.xhtml

Thanks

**Comment by elrob** [comment]
> > > Still having issues :( even with #152
> > 
> > 
> > @milktea02 What issues are you having? Are they related to truncation (this github issue tracks the truncation problem)?
> 
> @elrob Tried Clean Code (9780136083238) and still get truncation. I'm logging in via SSO if that might be the issue.

@milktea02 I have updated my change to restore the code that I thought was unnecessary. It was unnecessary for me but I'm not using SSO. Maybe you can try the latest version of my branch and see if it works for you now. I don't have SSO so I can't test it myself.

@AsimShakour Are you using SSO too? Maybe that's the problem. Can you also try with the latest change I have made (updated just now).

**Comment by varta2014** [comment]
**elrob** thank you  code work perfect !

**Comment by vikdean** [comment]
> > > > Still having issues :( even with #152
> > > 
> > > 
> > > @milktea02 What issues are you having? Are they related to truncation (this github issue tracks the truncation problem)?
> > 
> > 
> > @elrob Tried Clean Code (9780136083238) and still get truncation. I'm logging in via SSO if that might be the issue.
> 
> @milktea02 I have updated my change to restore the code that I thought was unnecessary. It was unnecessary for me but I'm not using SSO. Maybe you can try the latest version of my branch and see if it works for you now. I don't have SSO so I can't test it myself.
> 
> @AsimShakour Are you using SSO too? Maybe that's the problem. Can you also try with the latest change I have made (updated just now).

Just tested it with `9780135262047`; SSO works, but it still downloads the books partially.

**Comment by varta2014** [comment]
**elrob**
i test this 9781119558439
but not work just 17 pages :(

**Comment by brookscl** [comment]
For those of you still having trouble: delete the Books directory that is created for the downloads. Then retry your download. I found that the tool will not re-download chapters it thinks are already there. I was able to download book 9781119558439 without any problems. Not familiar, but it seemed complete.

**Comment by vikdean** [comment]
> For those of you still having trouble: delete the Books directory that is created for the downloads. Then retry your download. I found that the tool will not re-download chapters it thinks are already there. I was able to download book 9781119558439 without any problems. Not familiar, but it seemed complete.

Tried it 3 times in a row, issue is still the same for `9780135262047`

**Comment by varta2014** [comment]
i thinks some books we can't download also like 9781949991932

**Comment by mukuntharajaa** [comment]
> > For those of you still having trouble: delete the Books directory that is created for the downloads. Then retry your download. I found that the tool will not re-download chapters it thinks are already there. I was able to download book 9781119558439 without any problems. Not familiar, but it seemed complete.
> 
> Tried it 3 times in a row, issue is still the same for `9780135262047`

I have also tried downloading this ebook and accessed random pages, @elrob`s  fix is working fine.

**Comment by vikdean** [comment]
> > > For those of you still having trouble: delete the Books directory that is created for the downloads. Then retry your download. I found that the tool will not re-download chapters it thinks are already there. I was able to download book 9781119558439 without any problems. Not familiar, but it seemed complete.
> > 
> > 
> > Tried it 3 times in a row, issue is still the same for `9780135262047`
> 
> I have also tried downloading this ebook and accessed random pages, @elrob`s fix is working fine.

Check the Chapter beginnings... it only captures a couple of lines, the rest is truncated...
Also, whats the epub size for you? Mine is 3MB

**Comment by mukuntharajaa** [comment]
> > > > For those of you still having trouble: delete the Books directory that is created for the downloads. Then retry your download. I found that the tool will not re-download chapters it thinks are already there. I was able to download book 9781119558439 without any problems. Not familiar, but it seemed complete.
> > > 
> > > 
> > > Tried it 3 times in a row, issue is still the same for `9780135262047`
> > 
> > 
> > I have also tried downloading this ebook and accessed random pages, @elrob`s fix is working fine.
> 
> Check the Chapter beginnings... it only captures a couple of lines, the rest is truncated...
> Also, whats the epub size for you? Mine is 3MB

Its 114M MB. I have checked random chapters for its beginnings and its end. I do not see any truncation. Your case could be different. While downloading add "--preserve-log" and then check for anything reported in that. Create a new issue if required.

**Comment by vikdean** [comment]
> --preserve-log

I dont know whats going on, but I did a fresh install, and the issue still exist for me... the generated .epub is 3.4MB still.
Checked the log, its completely error-free. You can find it attached.
[log.txt](https://github.com/lorenzodifuccia/safaribooks/files/3845762/log.txt)

**Comment by elrob** [comment]
I think there is still an issue for some people. Perhaps only those that use SSO. I guess there is another cookie or more cookies that are missing. I can't test this because I don't know what cookies are missing because it works for me.

Another PR has been created which might solve the issue for some people: https://github.com/lorenzodifuccia/safaribooks/pull/153 @McPatate seems to have found another cookie that might be getting lost. Maybe try that version if you're still having issues.

**Comment by elrob** [comment]
Also, my safari books account has been 'canceled'. I'm trying to find out the cause. There's a chance it's related to using this tool. Therefore, people might want to use with caution.

**Comment by vikdean** [comment]
> I think there is still an issue for some people. Perhaps only those that use SSO. I guess there is another cookie or more cookies that are missing. I can't test this because I don't know what cookies are missing because it works for me.
> 
> Another PR has been created which might solve the issue for some people: #153 @McPatate seems to have found another cookie that might be getting lost. Maybe try that version if you're still having issues.

Tried that just now... same thing.

**Comment by McPatate** [comment]
> > I think there is still an issue for some people. Perhaps only those that use SSO. I guess there is another cookie or more cookies that are missing. I can't test this because I don't know what cookies are missing because it works for me.
> > Another PR has been created which might solve the issue for some people: #153 @McPatate seems to have found another cookie that might be getting lost. Maybe try that version if you're still having issues.
> 
> Tried that just now... same thing.

Do you have the epub's id so we can give it a go ourselves ?

**Comment by vikdean** [comment]
> > > I think there is still an issue for some people. Perhaps only those that use SSO. I guess there is another cookie or more cookies that are missing. I can't test this because I don't know what cookies are missing because it works for me.
> > > Another PR has been created which might solve the issue for some people: #153 @McPatate seems to have found another cookie that might be getting lost. Maybe try that version if you're still having issues.
> > 
> > 
> > Tried that just now... same thing.
> 
> Do you have the epub's id so we can give it a go ourselves ?

You mean the book id? Its 9780135262047

**Comment by McPatate** [comment]
Indeed, it doesn't work. I'm looking into what could be the problem. @elrob have you tried with that book id?

**Comment by varta2014** [comment]
**elrob** code not work please fix
thank you

**Comment by McPatate** [comment]
> > > > I think there is still an issue for some people. Perhaps only those that use SSO. I guess there is another cookie or more cookies that are missing. I can't test this because I don't know what cookies are missing because it works for me.
> > > > Another PR has been created which might solve the issue for some people: #153 @McPatate seems to have found another cookie that might be getting lost. Maybe try that version if you're still having issues.
> > > 
> > > 
> > > Tried that just now... same thing.
> > 
> > 
> > Do you have the epub's id so we can give it a go ourselves ?
> 
> You mean the book id? Its 9780135262047

It works with my code : https://github.com/McPatate/orly_book_extractor. The only problem is that it's pretty ugly 😓

**Comment by varta2014** [comment]
**McPatate** yes code work thank you
but you need fix bookmark and some error ... we wait for final code thanks

**Comment by elrob** [comment]
@varta2014 @vikdean 
I have investigated further and made some more changes that resolve another couple of issues with login.
Can you try again with a fresh version of this: https://github.com/elrob/safaribooks
Make sure to delete anything in the `Books` directory before trying.

**Comment by vikdean** [comment]
> @varta2014 @vikdean
> I have investigated further and made some more changes that resolve another couple of issues with login.
> Can you try again with a fresh version of this: https://github.com/elrob/safaribooks
> Make sure to delete anything in the `Books` directory before trying.

I've just tried it; SSO authentication completely broken... it does not work at all. The only thing I get is this:
```
[18/Nov/2019 11:43:28] ** Welcome to SafariBooks! **
[18/Nov/2019 11:43:30] Authentication issue: unable to access profile page.
[18/Nov/2019 11:43:30] Last request done:
	URL: https://learning.oreilly.com/profile/
	DATA: None
	OTHERS: {}

	307
	server: istio-envoy
	cache-control: max-age=0
	content-type: text/plain; charset=utf-8
	location: /accounts/login/?next=%2Fprofile%2F
	x-envoy-upstream-service-time: 814
	x-powered-by: Express
	Accept-Ranges: bytes, bytes
	Content-Length: 70
	Date: Mon, 18 Nov 2019 10:43:30 GMT
	Via: 1.1 varnish
	Connection: keep-alive
	X-Client-IP: 188.143.125.75
	X-Served-By: cache-lcy19235-LCY
	X-Cache: MISS
	X-Cache-Hits: 0
	X-Timer: S1574073809.109248,VS0,VE945
	Vary: Accept,Accept, Accept-Encoding, Authorization, Cookie

Temporary Redirect. Redirecting to /accounts/login/?next=%2Fprofile%2F
```

**Comment by elrob** [comment]
@vikdean 
How are you attempting to use the script with SSO? I don't think this script will support SSO or ever has, except if you provide your own `cookies.json`. One of the recent changes I have made is to confirm the login before continuing with processing the book. Previously, the book processing would continue but then you'd just get a book with truncated chapters because the login had failed. Now it fails faster if there is an issue.

If you want to use the script with SSO, I think you need to do the following (I'll provide firefox instructions but it should also be possible with other browsers):
1. Login in the browser as you normally would: https://learning.oreilly.com
2. Access the profile page: https://learning.oreilly.com/profile/
3. Open the developer tools: Press `F12` in firefox
4. At the bottom there is a console where you can type commands. Paste the following in there (the first time you do this it may ask you to `allow pasting`):
    `var output = {};document.cookie.split(/\s*;\s*/).forEach(function(pair) {pair = pair.split(/\s*=\s*/);output[pair[0]] = pair.splice(1).join('=');});console.log(JSON.stringify(output));` 
(Credit to https://github.com/lorenzodifuccia/safaribooks/issues/2#issuecomment-429343521)
5. Copy the JSON output and save it in a file called `cookies.json` in the same directory as the safaribooks code.
6. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`

**Comment by gattytto** [comment]
kampyleSessionPageCounter

**Comment by vikdean** [comment]
>     6\. python3 safaribooks.py 9780135262047

Yes,

> @vikdean
> How are you attempting to use the script with SSO? I don't think this script will support SSO or ever has, except if you provide your own `cookies.json`. One of the recent changes I have made is to confirm the login before continuing with processing the book. Previously, the book processing would continue but then you'd just get a book with truncated chapters because the login had failed. Now it fails faster if there is an issue.
> 
> If you want to use the script with SSO, I think you need to do the following (I'll provide firefox instructions but it should also be possible with other browsers):
> 
>     1. Login in the browser as you normally would: https://learning.oreilly.com
> 
>     2. Access the profile page: https://learning.oreilly.com/profile/
> 
>     3. Open the developer tools: Press `F12` in firefox
> 
>     4. At the bottom there is a console where you can type commands. Paste the following in there (the first time you do this it may ask you to `allow pasting`):
>        `var output = {};document.cookie.split(/\s*;\s*/).forEach(function(pair) {pair = pair.split(/\s*=\s*/);output[pair[0]] = pair.splice(1).join('=');});console.log(JSON.stringify(output));`
>        (Credit to [#2 (comment)](https://github.com/lorenzodifuccia/safaribooks/issues/2#issuecomment-429343521))
> 
>     5. Copy the JSON output and save it in a file called `cookies.json` in the same directory as the safaribooks code.
> 
>     6. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`

Yes, that's exactly how I'm using it, right to the dot.
Error message is still the same, no matter what I do:
```
[19/Nov/2019 09:05:26] ** Welcome to SafariBooks! **
[19/Nov/2019 09:05:27] Authentication issue: unable to access profile page.
[19/Nov/2019 09:05:27] Last request done:
        URL: https://learning.oreilly.com/profile/
        DATA: None
        OTHERS: {}

        307
        Connection: keep-alive
        Content-Length: 70
        server: istio-envoy
        cache-control: max-age=0
        content-type: text/plain; charset=utf-8
        location: /accounts/login/?next=%2Fprofile%2F
        x-envoy-upstream-service-time: 140
        x-powered-by: Express
        Accept-Ranges: bytes, bytes
        Date: Tue, 19 Nov 2019 08:05:27 GMT
        Via: 1.1 varnish
        X-Client-IP: 195.212.29.169
        X-Served-By: cache-lga21943-LGA
        X-Cache: MISS
        X-Cache-Hits: 0
        X-Timer: S1574150727.907460,VS0,VE422
        Vary: Accept,Accept, Accept-Encoding, Authorization, Cookie

Temporary Redirect. Redirecting to /accounts/login/?next=%2Fprofile%2F
```

**Comment by elrob** [comment]
@vikdean 
I think I've found the problem.
Using `document.cookie` from the console does not include the `HttpOnly` cookies and they are definitely required.
I can't work out how to access these via the console but I was able to find a way to get them that isn't too painful.
1. Login as usual to https://learning.oreilly.com/
2. Open the developer tools with `F12`
3. Go to `Network` tab in the developer tools
4. Access the profile page in the browser: https://learning.oreilly.com/profile/
5. In the `Network` tab, click on the request to `/profile/` (it should be the first one)
6. Click on the `Cookies` tab in the request information
7. Right-click on the `Request cookies` text and choose `Copy All`
8. Paste this into the `cookies.json` file and then remove the outer section of the JSON document
9. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`

p.s. `sudo` is not necessary.

**Comment by vikdean** [comment]
> @vikdean
> I think I've found the problem.
> Using `document.cookie` from the console does not include the `HttpOnly` cookies and they are definitely required.
> I can't work out how to access these via the console but I was able to find a way to get them that isn't too painful.
> 
>     1. Login as usual to https://learning.oreilly.com/
> 
>     2. Open the developer tools with `F12`
> 
>     3. Go to `Network` tab in the developer tools
> 
>     4. Access the profile page in the browser: https://learning.oreilly.com/profile/
> 
>     5. In the `Network` tab, click on the request to `/profile/` (it should be the first one)
> 
>     6. Click on the `Cookies` tab in the request information
> 
>     7. Right-click on the `Request cookies` text and choose `Copy All`
> 
>     8. Paste this into the `cookies.json` file and then remove the outer section of the JSON document
> 
>     9. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`
> 
> 
> p.s. `sudo` is not necessary.

Yes!!! Its working now, thanks a lot!

**Comment by varta2014** [comment]
**elrob** code not work again :( we get not full ebook

**Comment by varta2014** [comment]
**elrob** hi my friend ca n help here please thank you
https://github.com/lmbringas/packtpub-downloader/issues/25

**Comment by elrob** [comment]
> **elrob** hi my friend can help here please thank you
> [lmbringas/packtpub-downloader#25](https://github.com/lmbringas/packtpub-downloader/issues/25)

You're requesting me to help on another project? I'm not sure that belongs in this issue thread. 
I'm not willing to help on that other project.

**Comment by lorenzodifuccia** [comment]
I pushed some changes, try with the last commit...

Thank you @elrob for your great job. 🎉 
Cheers 🍺

**Comment by lorenzodifuccia** [comment]
News???

**Comment by nkkarthik** [comment]
Latest is working for me (verified with couple of books, that were being truncated before this push).  Thank you @elrob @lorenzodefuccia

**Comment by mukuntharajaa** [comment]
> News???

Checked with original book id and some new books. Working perfectly. Thanks.

**Comment by roleed** [comment]
Have the most recent commit, seems to get most books fine and haven't encountered many errors but there is a grey background to all books that never seemed to happen before. Also for coding books --no-kindle used to remove the scrollbar most of the time.

**Comment by villancikos** [comment]
> @vikdean
> I think I've found the problem.
> Using `document.cookie` from the console does not include the `HttpOnly` cookies and they are definitely required.
> I can't work out how to access these via the console but I was able to find a way to get them that isn't too painful.
> 
> 1. Login as usual to https://learning.oreilly.com/
> 2. Open the developer tools with `F12`
> 3. Go to `Network` tab in the developer tools
> 4. Access the profile page in the browser: https://learning.oreilly.com/profile/
> 5. In the `Network` tab, click on the request to `/profile/` (it should be the first one)
> 6. Click on the `Cookies` tab in the request information
> 7. Right-click on the `Request cookies` text and choose `Copy All`
> 8. Paste this into the `cookies.json` file and then remove the outer section of the JSON document
> 9. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`
> 
> p.s. `sudo` is not necessary.

I think something is not right or just changed. I tried but this repo master and yours @elrob with no success. The Developer Tools Network tab inside the cookies section (profile page) won’t show any httpOnly cookie. Don’t know if this is just me.

**Comment by elrob** [comment]
> > @vikdean
> > I think I've found the problem.
> > Using `document.cookie` from the console does not include the `HttpOnly` cookies and they are definitely required.
> > I can't work out how to access these via the console but I was able to find a way to get them that isn't too painful.
> > 
> > 1. Login as usual to https://learning.oreilly.com/
> > 2. Open the developer tools with `F12`
> > 3. Go to `Network` tab in the developer tools
> > 4. Access the profile page in the browser: https://learning.oreilly.com/profile/
> > 5. In the `Network` tab, click on the request to `/profile/` (it should be the first one)
> > 6. Click on the `Cookies` tab in the request information
> > 7. Right-click on the `Request cookies` text and choose `Copy All`
> > 8. Paste this into the `cookies.json` file and then remove the outer section of the JSON document
> > 9. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`
> > 
> > p.s. `sudo` is not necessary.
> 
> I think something is not right or just changed. I tried but this repo master and yours @elrob with no success. The Developer Tools Network tab inside the cookies section (profile page) won’t show any httpOnly cookie. Don’t know if this is just me.

@villancikos 
I have just tested again with my fork of the repo. It is working fine for me when I download the cookies following the instructions above. `httpOnly` doesn't refer to the name of a cookie. Two of the cookies `groot_sessionid` and `orm-rt` are set to `httpOnly=true` so it means some other methods of downloading the cookies don't work. The method above does work for me in firefox. If you're still having an issue, can you explain where it is going wrong and I might be able to help.

**Comment by villancikos** [comment]
> > > @vikdean
> > > I think I've found the problem.
> > > Using `document.cookie` from the console does not include the `HttpOnly` cookies and they are definitely required.
> > > I can't work out how to access these via the console but I was able to find a way to get them that isn't too painful.
> > > 
> > > 1. Login as usual to https://learning.oreilly.com/
> > > 2. Open the developer tools with `F12`
> > > 3. Go to `Network` tab in the developer tools
> > > 4. Access the profile page in the browser: https://learning.oreilly.com/profile/
> > > 5. In the `Network` tab, click on the request to `/profile/` (it should be the first one)
> > > 6. Click on the `Cookies` tab in the request information
> > > 7. Right-click on the `Request cookies` text and choose `Copy All`
> > > 8. Paste this into the `cookies.json` file and then remove the outer section of the JSON document
> > > 9. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`
> > > 
> > > p.s. `sudo` is not necessary.
> > 
> > 
> > I think something is not right or just changed. I tried but this repo master and yours @elrob with no success. The Developer Tools Network tab inside the cookies section (profile page) won’t show any httpOnly cookie. Don’t know if this is just me.
> 
> @villancikos
> I have just tested again with my fork of the repo. It is working fine for me when I download the cookies following the instructions above. `httpOnly` doesn't refer to the name of a cookie. Two of the cookies `groot_sessionid` and `orm-rt` are set to `httpOnly=true` so it means some other methods of downloading the cookies don't work. The method above does work for me in firefox. If you're still having an issue, can you explain where it is going wrong and I might be able to help.

Hi @elrob . Thanks for your answer. First, I know httpOnly is a type of cookie. It is strange that my Chrome does not "tick" them in the dev tools. 
I tried adding the common cookies using the javascript output and manually these two cookies: `groot_sessionid` and `orm-rt` but I am still getting a truncated epub. 
BTW I am using your repo on the master branch. 
Just for reference, the book id is `9781491973783`.

**Comment by elrob** [comment]
Hi again @villancikos ,
I have tested again and I tested with the specific book id `9781491973783`. I did not have any issues following the steps I have listed and using firefox. All chapters seem to be complete. I can't see any truncation.
Maybe you are still missing some cookies or something.
My cookies.json file looks like (values replaced with `XXX`):
```JSON
{
  "BrowserCookie": "XXX",
  "_gcl_au": "XXX",
  "cd_user_id": "XXX",
  "csrfsafari": "XXX",
  "groot_sessionid": "XXX",
  "kampyleSessionPageCounter": "XXX",
  "kampyleUserSession": "XXX",
  "kampyleUserSessionsCount": "XXX",
  "kampyle_userid": "XXX",
  "logged_in": "XXX",
  "orm-jwt": "XXX",
  "orm-rt": "XXX",
  "salesforce_id": "XXX",
  "sessionid": "XXX"
}
```

**Comment by mukuntharajaa** [comment]
I am facing new issue with book "9781492047834". Should I file new bug or part of this ?

[-] Downloading book contents... (16 chapters)                                                                                      
    [#########################################################################################################################] 100%
[-] Downloading book CSSs... (4 files)                                                                                              
[#] HTTPSConnectionPool(host='learning.oreilly.com', port=443): Max retries exceeded with url: /static/CACHE/css/output.68851547a55f.css (Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:2508)')))
[#] HTTPSConnectionPool(host='learning.oreilly.com', port=443): Max retries exceeded with url: /static/css/annotator.e3b0c44298fc.css (Caused by SSLError(SSLError(1, '[SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:2508)')))
[#] Error trying to retrieve this CSS: /home/xxxyyyyxxx/safaribooks/Books/Monolith to Microservices (9781492047834)/OEBPS/Styles/Style01.css
    From: https://learning.oreilly.com/static/CACHE/css/output.68851547a55f.css
Process Process-2:
[#] Error trying to retrieve this CSS: /home/xxxyyyyxxx/safaribooks/Books/Monolith to Microservices (9781492047834)/OEBPS/Styles/Style02.css
    From: https://learning.oreilly.com/static/css/annotator.e3b0c44298fc.css
Process Process-3:
Traceback (most recent call last):
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "safaribooks.py", line 829, in _thread_download_css
    s.write(response.content)
AttributeError: 'int' object has no attribute 'content'
Traceback (most recent call last):
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "safaribooks.py", line 829, in _thread_download_css
    s.write(response.content)
AttributeError: 'int' object has no attribute 'content'
[-] Downloading book images... (132 files)                                                                                          
    [#########################################################################################################################] 100%
[-] Creating EPUB file...                                                                                                           
[*] Done: /home/xxxyyyyxxx/safaribooks/Books/Monolith to Microservices (9781492047834)/9781492047834.epub                 

    If you like it, please * this project on GitHub to make it known:
        https://github.com/lorenzodifuccia/safaribooks
    e don't forget to renew your Safari Books Online subscription:
        https://learning.oreilly.com

[!] Bye!!

**Comment by villancikos** [comment]
> Hi again @villancikos ,
> I have tested again and I tested with the specific book id `9781491973783`. I did not have any issues following the steps I have listed and using firefox. All chapters seem to be complete. I can't see any truncation.
> Maybe you are still missing some cookies or something.
> My cookies.json file looks like (values replaced with `XXX`):
> 
> ```json
> {
>   "BrowserCookie": "XXX",
>   "_gcl_au": "XXX",
>   "cd_user_id": "XXX",
>   "csrfsafari": "XXX",
>   "groot_sessionid": "XXX",
>   "kampyleSessionPageCounter": "XXX",
>   "kampyleUserSession": "XXX",
>   "kampyleUserSessionsCount": "XXX",
>   "kampyle_userid": "XXX",
>   "logged_in": "XXX",
>   "orm-jwt": "XXX",
>   "orm-rt": "XXX",
>   "salesforce_id": "XXX",
>   "sessionid": "XXX"
> }
> ```

Very strange. My cookies are as follows (still truncated, missing images, etc).

`{
    "BrowserCookie": "xxxxx,
    "_fbp": "xxxxx",
    "_ga": "xxxxx",
    "_gat_UA-112091926-1": "xxxxx",
    "_gcl_au": "xxxxx",
    "_gid": "xxxxx",
    "_mkto_trk": "xxxxx,
    "_vwo_uuid_v2": "xxxxx",
    "cd_user_id": "xxxxx",
    "csrfsafari": "xxxxx",
    "groot_sessionid": "xxxxx",
    "kampyleSessionPageCounter": "xxxxx",
    "kampyleUserSession": "xxxxx",
    "kampyleUserSessionsCount": "xxxxx",
    "kampyle_userid": "xxxxx",
    "logged_in": "xxxxx",
    "orm-jwt": "xxxxx",
    "orm-rt": "xxxxx",
    "recently-viewed": "xxxxx",
    "sessionid": "xxxxx"

}`

**Comment by roleed** [comment]
> Hi again @villancikos ,
> I have tested again and I tested with the specific book id `9781491973783`. I did not have any issues following the steps I have listed and using firefox. All chapters seem to be complete. I can't see any truncation.
> Maybe you are still missing some cookies or something.
> My cookies.json file looks like (values replaced with `XXX`):
> 
> ```json
> {
>   "BrowserCookie": "XXX",
>   "_gcl_au": "XXX",
>   "cd_user_id": "XXX",
>   "csrfsafari": "XXX",
>   "groot_sessionid": "XXX",
>   "kampyleSessionPageCounter": "XXX",
>   "kampyleUserSession": "XXX",
>   "kampyleUserSessionsCount": "XXX",
>   "kampyle_userid": "XXX",
>   "logged_in": "XXX",
>   "orm-jwt": "XXX",
>   "orm-rt": "XXX",
>   "salesforce_id": "XXX",
>   "sessionid": "XXX"
> }
> ```

Hi,

What method did you use to extract the cookies.json in FireFox? Many thanks!

**Comment by villancikos** [comment]
> > Hi again @villancikos ,
> > I have tested again and I tested with the specific book id `9781491973783`. I did not have any issues following the steps I have listed and using firefox. All chapters seem to be complete. I can't see any truncation.
> > Maybe you are still missing some cookies or something.
> > My cookies.json file looks like (values replaced with `XXX`):
> > ```json
> > {
> >   "BrowserCookie": "XXX",
> >   "_gcl_au": "XXX",
> >   "cd_user_id": "XXX",
> >   "csrfsafari": "XXX",
> >   "groot_sessionid": "XXX",
> >   "kampyleSessionPageCounter": "XXX",
> >   "kampyleUserSession": "XXX",
> >   "kampyleUserSessionsCount": "XXX",
> >   "kampyle_userid": "XXX",
> >   "logged_in": "XXX",
> >   "orm-jwt": "XXX",
> >   "orm-rt": "XXX",
> >   "salesforce_id": "XXX",
> >   "sessionid": "XXX"
> > }
> > ```
> 
> Hi,
> 
> What method did you use to extract the cookies.json in FireFox? Many thanks!

As previous messages inside this thread suggest:

1. Open Firefox.
2. Open Developer tools (alt+command+I in Mac, control+shift+I Win).
3. Navigate to /profile/ URL,.
4. At the Network Tab select /profile/.
5. Go to the Cookies "sub-tab".
6. Right Click at the "Request Cookies" headline and click on COPY ALL.
7. Paste them to a Cookies.json file.

**Comment by AsimShakour** [comment]
Thanks all, I will look into when I need to download another book. Still got to finish the ones I already downloaded :P

**Comment by munish259272** [comment]
another way is 
https://stackoverflow.com/questions/1324421/how-to-get-past-the-login-page-with-wget#answer-37780143

**Comment by azmatsiddique** [comment]
please provide images or video to get cookies.json file from inpection in mac

**Comment by obar1** [comment]
@azmatsiddique ahhaha are you serious

**Comment by darshanmnyk** [comment]
Thanks a lot guys! Works splendidly.

**Comment by MuhammedElGanzory** [comment]
> Thanks a lot guys! Works splendidly.

can you help me I'm trying to download also !!

**Comment by MuhammedElGanzory** [comment]
how to solve !!
Traceback (most recent call last):
  File "C:\Users\LuckyMoon\Downloads\safarinew\safaribooks.py", line 10, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'

**Comment by dan-r95** [comment]
The only cookie which you really needs is the `orm-jwt` I think.  
![image](https://user-images.githubusercontent.com/18677643/107401633-5d572c00-6b03-11eb-8f20-0d4b5798b74c.png)

**Comment by EmanuelMtzV** [comment]
> @azmatsiddique ahhaha are you serious

i cant get the cookies.json file either. Any clue where to get it ?

**Comment by akriaueno** [comment]
Below script works for me.
Paste this script into console of Chrome DevTools and get cookies. 
```js
console.log(JSON.stringify(document.cookie.split(';').map(c => c.split('=')).map(i => [i[0].trim(), i[1].trim()]).reduce((r, i) => {r[i[0]] = i[1]; return r;}, {})))
```

**Comment by venkatasrinivasnamburi** [comment]
i see this error on windows ( while trying to download a book) , any inputs ?

Traceback (most recent call last):
  File "safaribooks.py", line 16, in <module>
    from lxml import html, etree
ModuleNotFoundError: No module named 'lxml'

**Comment by munish259272** [comment]
> > > > @vikdean
> > > > I think I've found the problem.
> > > > Using `document.cookie` from the console does not include the `HttpOnly` cookies and they are definitely required.
> > > > I can't work out how to access these via the console but I was able to find a way to get them that isn't too painful.
> > > > 
> > > > 1. Login as usual to https://learning.oreilly.com/
> > > > 2. Open the developer tools with `F12`
> > > > 3. Go to `Network` tab in the developer tools
> > > > 4. Access the profile page in the browser: https://learning.oreilly.com/profile/
> > > > 5. In the `Network` tab, click on the request to `/profile/` (it should be the first one)
> > > > 6. Click on the `Cookies` tab in the request information
> > > > 7. Right-click on the `Request cookies` text and choose `Copy All`
> > > > 8. Paste this into the `cookies.json` file and then remove the outer section of the JSON document
> > > > 9. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`
> > > > 
> > > > p.s. `sudo` is not necessary.
> > > 
> > > 
> > > I think something is not right or just changed. I tried but this repo master and yours @elrob with no success. The Developer Tools Network tab inside the cookies section (profile page) won’t show any httpOnly cookie. Don’t know if this is just me.
> > 
> > 
> > @villancikos
> > I have just tested again with my fork of the repo. It is working fine for me when I download the cookies following the instructions above. `httpOnly` doesn't refer to the name of a cookie. Two of the cookies `groot_sessionid` and `orm-rt` are set to `httpOnly=true` so it means some other methods of downloading the cookies don't work. The method above does work for me in firefox. If you're still having an issue, can you explain where it is going wrong and I might be able to help.
> 
> Hi @elrob . Thanks for your answer. First, I know httpOnly is a type of cookie. It is strange that my Chrome does not "tick" them in the dev tools.
> I tried adding the common cookies using the javascript output and manually these two cookies: `groot_sessionid` and `orm-rt` but I am still getting a truncated epub.
> BTW I am using your repo on the master branch.
> Just for reference, the book id is `9781491973783`.

This does not work anymore

**Comment by domrany64** [comment]
I could use it perfectly fine, after trouble I had to obtain the cookies in the right structure.
To get the cookies, I'm using a Chrome extension called `Cookie-Editor`.

1. Open the O'Reilly website in the Chrome and log in using SSO
2. Open `Cookie-Editor`.
3. Click `EXPORT` which is the right icon in the bottom of `Cookie-Editor`'s window.
4. Paste the cookies (which are in the clipboard, now) to an editor.
5. Find the `"name": "orm-jwt",` among the text and copy the `value` from that section
6. Create the `cookies.json` file  like this `{"orm-jwt": "XXX"}` where XXX is copied value from step 5
7. Run `python3 safaribooks.py XXXXXXXXXXX` and enjoy the EPUB book.

**Comment by Marakai** [comment]
> Below script works for me. Paste this script into console of Chrome DevTools and get cookies.
> 
> ```js
> console.log(JSON.stringify(document.cookie.split(';').map(c => c.split('=')).map(i => [i[0].trim(), i[1].trim()]).reduce((r, i) => {r[i[0]] = i[1]; return r;}, {})))
> ```

Using this with Chrome in mid 2022 and it seems to be the easiest approach by far. Copy-pasted into `cookies.json` in the script directory and it works like a charm!

**Comment by MrDandas** [comment]
might add PR later, but for now, I've provided my own solution based on `browser_cookie3`. I just log in through browser and let library to grab cookies from browser:


```
#            self.session.cookies.update(json.load(open(COOKIES_FILE)))
            self.session.cookies = browser_cookie3.firefox(domain_name='oreilly.com')
```

**Comment by Metal-Milonga** [comment]
> > Below script works for me. Paste this script into console of Chrome DevTools and get cookies.
> > ```js
> > console.log(JSON.stringify(document.cookie.split(';').map(c => c.split('=')).map(i => [i[0].trim(), i[1].trim()]).reduce((r, i) => {r[i[0]] = i[1]; return r;}, {})))
> > ```
> 
> Using this with Chrome in mid 2022 and it seems to be the easiest approach by far. Copy-pasted into `cookies.json` in the script directory and it works like a charm!

worked again with Chrome in July 2023.  Copy the returned json to cookies.json, and download was successful.

**Comment by KonScanner** [comment]
@domrany64's answer still works!

**Comment by eurubkov** [comment]
I am not finding the "name": "orm-jwt", or just "orm-jwt" in the cookies at all. Is it still working for others?

**Comment by eurubkov** [comment]
Looks like it's the Cookie-Editor extension that didn't get all the cookies. 
Here's how I got it instead:
Right-Click -> Inspect Element -> Application -> Cookies -> selecting the orelly website -> orm-jwt and copying the value from there.
Then created the cookies.json as mentioned above and it worked.

**Comment by yuletide** [comment]
This method generally worked for me per @albertocavalcante's post

**Comment by henryheim** [comment]
Like some others, I have an SSO login to O'Reilly and had trouble with the script as it is in the master branch. @albertocavalcante 's method, combined with the dev tools one-liner by @akriaueno fixed everything and the tool works easily now. Just save the output of the dev tools script into `cookies.json` in the safaribooks directory and run the script without any auth parameters.

Thank you both for your debugging and guides.

**Comment by tuha1994** [comment]
> > @azmatsiddique ahhaha are you serious
> 
> i cant get the cookies.json file either. Any clue where to get it ?

create a file 'cookies.json' in the same folder as safaribooks, then open F12 in the sonsole tab and paste the following string to get cookies, copy all the returned values ​​and save them to cookies.json. then run the command 'python safaribooks.py xxxxx' where xxxxx is your document code

**Comment by jjmca** [comment]
For people hitting the error `[#] Authentication issue: unable to access profile page.` error, I had to make a modification to allow redirects. Changing `allow_redirects=True` [here](https://github.com/lorenzodifuccia/safaribooks/blob/master/safaribooks.py#L428) solved my issue and worked flawlessly. Great tool, thank you!

```python
response = getattr(self.session, "post" if is_post else "get")(
    url,
    data=data,
    allow_redirects=True,
    **kwargs
)
```

**Comment by bakudas** [comment]
> > > Below script works for me. Paste this script into console of Chrome DevTools and get cookies.
> > > console.log(JSON.stringify(document.cookie.split(';').map(c => c.split('=')).map(i => [i[0].trim(), i[1].trim()]).reduce((r, i) => {r[i[0]] = i[1]; return r;}, {})))
> > 
> > 
> > Using this with Chrome in mid 2022 and it seems to be the easiest approach by far. Copy-pasted into `cookies.json` in the script directory and it works like a charm!
> 
> worked again with Chrome in July 2023. Copy the returned json to cookies.json, and download was successful.

still working in march 2025. simple and precise! Thanks and happy studying!

**Comment by munish259272** [comment]
```python
from pycookiecheat import BrowserType, firefox_cookies
import requests, json
import argparse
import sys

# Set up argument parser
parser = argparse.ArgumentParser(description='Get cookie file for a website')
parser.add_argument('url', type=str, nargs='?', default='https://learning.oreilly.com/profile/', help='Cookie URL (default: https://learning.oreilly.com/profile/)')

try:
    args = parser.parse_args()
except SystemExit:
    # Print a custom usage message if parsing fails
    print("Error: Incorrect arguments provided.")
    parser.print_usage()
    sys.exit(1)

# Uses Chrome's default cookies filepath by default
#cookies = firefox_cookies(url)
#r = requests.get(args.url, cookies=cookies)

# Using an alternate browser
cookies = firefox_cookies(args.url, browser=BrowserType.FIREFOX)
r = requests.get(args.url, cookies=cookies)

# Convert the cookies dictionary to a JSON string
cookies_json = json.dumps(cookies, indent=4)

# Save cookies JSON to a file
with open('cookies.json', 'w') as f:
    f.write(cookies_json)
    
# Print the JSON-formatted cookies
print(cookies_json)
```

https://github.com/n8henrie/pycookiecheat

This works for me to get the cookies once i login to firefox

**Comment by guillaumehx** [comment]
It does not seem to work for me. With the `--cred` option I get  "`[-] Logging into Safari Books Online...`" forever.
Fetching the cookies with the previous answers works but once I run the command without `--cred` it prints out : 
```
[-] Successfully authenticated.
[*] Retrieving book info...
[#] API: [+] Use the `--cred` or `--login` options in order to perform the auth login to Safari.
[!] Aborting...
```

**Comment by guillaumehx** [comment]
> It does not seem to work for me. With the `--cred` option I get "`[-] Logging into Safari Books Online...`" forever. Fetching the cookies with the previous answers works but once I run the command without `--cred` it prints out :
> 
> ```
> [-] Successfully authenticated.
> [*] Retrieving book info...
> [#] API: [+] Use the `--cred` or `--login` options in order to perform the auth login to Safari.
> [!] Aborting...
> ```

Fixed it, in fact I was using a wrong book id (the book from the example above does not work anymore `9780135262047`)

**Comment by karthikdm91** [comment]
> > It does not seem to work for me. With the `--cred` option I get "`[-] Logging into Safari Books Online...`" forever. Fetching the cookies with the previous answers works but once I run the command without `--cred` it prints out :
> > ```
> > [-] Successfully authenticated.
> > [*] Retrieving book info...
> > [#] API: [+] Use the `--cred` or `--login` options in order to perform the auth login to Safari.
> > [!] Aborting...
> > ```
> 
> Fixed it, in fact I was using a wrong book id (the book from the example above does not work anymore `9780135262047`)

Hi - Were you able to download complete book. I am able to download the book , but for some reason only first 4 pages of the  books are downloaded. I tried with multiple books and it does the same for all books. Am I missing something, Please advise.

**Comment by guillaumehx** [comment]
> > > It does not seem to work for me. With the `--cred` option I get "`[-] Logging into Safari Books Online...`" forever. Fetching the cookies with the previous answers works but once I run the command without `--cred` it prints out :
> > > ```
> > > [-] Successfully authenticated.
> > > [*] Retrieving book info...
> > > [#] API: [+] Use the `--cred` or `--login` options in order to perform the auth login to Safari.
> > > [!] Aborting...
> > > ```
> > 
> > 
> > Fixed it, in fact I was using a wrong book id (the book from the example above does not work anymore `9780135262047`)
> 
> Hi - Were you able to download complete book. I am able to download the book , but for some reason only first 4 pages of the books are downloaded. I tried with multiple books and it does the same for all books. Am I missing something, Please advise.

Yes I was able to download. You must have a valid license to O'Reilly (you have the first few pages because you can't see the complete book online either) then you just have to copy the cookies in a file called cookies.json then use `python3 safaribooks.py <BOOK_ID>`

**Comment by mniak** [comment]
I'm using the following
1. Use the browser extension https://cookie-editor.com/ (or any other that can output JSON)
2. Export the cookies of https://learning.oreilly.com as JSON to the clipboard
![Image](https://github.com/user-attachments/assets/afc73bcb-5898-4cd1-a565-34cb3761ff06)
3. Transform this into the expected `cookies.json` file:
```bash
pbpaste | jq 'map({(.name): .value}) | add' > cookies.json
```
4. Use the app without passing credentials.
```

**Comment by hohherndernharng** [comment]
> I'm using the following
> 
> 1. Use the browser extension https://cookie-editor.com/ (or any other that can output JSON)
> 2. Export the cookies of https://learning.oreilly.com as JSON to the clipboard
>    ![Image](https://github.com/user-attachments/assets/afc73bcb-5898-4cd1-a565-34cb3761ff06)
> 3. Transform this into the expected `cookies.json` file:
> 
> pbpaste | jq 'map({(.name): .value}) | add' > cookies.json
> 4. Use the app without passing credentials.

Thank you, its work.

**Comment by SimonPrato11** [comment]
Thank you, it works great!

**Comment by paciox** [comment]
> [@vikdean](https://github.com/vikdean) How are you attempting to use the script with SSO? I don't think this script will support SSO or ever has, except if you provide your own `cookies.json`. One of the recent changes I have made is to confirm the login before continuing with processing the book. Previously, the book processing would continue but then you'd just get a book with truncated chapters because the login had failed. Now it fails faster if there is an issue.
> 
> If you want to use the script with SSO, I think you need to do the following (I'll provide firefox instructions but it should also be possible with other browsers):
> 
> 1. Login in the browser as you normally would: https://learning.oreilly.com
> 2. Access the profile page: https://learning.oreilly.com/profile/
> 3. Open the developer tools: Press `F12` in firefox
> 4. At the bottom there is a console where you can type commands. Paste the following in there (the first time you do this it may ask you to `allow pasting`):
>    `var output = {};document.cookie.split(/\s*;\s*/).forEach(function(pair) {pair = pair.split(/\s*=\s*/);output[pair[0]] = pair.splice(1).join('=');});console.log(JSON.stringify(output));`
>    (Credit to [Cookies.json format #2 (comment)](https://github.com/lorenzodifuccia/safaribooks/issues/2#issuecomment-429343521))
> 5. Copy the JSON output and save it in a file called `cookies.json` in the same directory as the safaribooks code.
> 6. Run the script without passing credentials: `python3 safaribooks.py 9780135262047`

Worked by staying on https://learning.oreilly.com/ and using the quoted post javascript as of october 2025

**Comment by Aries1304** [comment]
<img width="1307" height="229" alt="Image" src="https://github.com/user-attachments/assets/fbb1ca56-5eb6-4350-b000-6826b8b3311f" />

**Comment by paciox** [comment]
> <img alt="Image" width="1307" height="229" src="https://private-user-images.githubusercontent.com/73518027/630435415-fbb1ca56-5eb6-4350-b000-6826b8b3311f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODU4MzM0NjQsIm5iZiI6MTc4NTgzMzE2NCwicGF0aCI6Ii83MzUxODAyNy82MzA0MzU0MTUtZmJiMWNhNTYtNWViNi00MzUwLWIwMDAtNjgyNmI4YjMzMTFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODA0VDA4NDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiM2UyMDFhZWMyYjEyOWU0NTQyNjI1YmRjNDNjODgyMGU5MzA3N2FlMTJkNGY5NjMwMThhMzVhYWVmYmIyNTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.od1j-GUZk-5qeXjGKjhN99QjbkpLpyps4DAJLsT5o-k">

just head to my github and clone the new version we made thanks to the help of @tuha1994 i fixed all I could and worked charms for me
